OWML.ModHelper.Input/ModInputHandler.cs

```diff
@@ -0,0 +1,387 @@
+                hash = -hash;
+            }
+            if (!_comboRegistry.ContainsKey(hash))
+            {
+                return null;
+            }
+
+            var combination = _comboRegistry[hash];
+            if (!(combination == _currentCombination) && countdownTrigger)
+            {
+                return null;
+            }
+
+            if (hash < MaxUsefulKey)
+            {
+                return combination;
+            }
+            while (hash > 0)
+            {
+                _timeout[hash % MaxUsefulKey] = Time.realtimeSinceStartup;
+                hash /= MaxUsefulKey;
+            }
+            return combination;
+        }
+
+        private void UpdateCurrentCombination()
+        {
+            if (Time.realtimeSinceStartup - _lastCombinationUpdate < Time.unscaledDeltaTime / 2)
+            {
+                return;
+            }
+            _lastCombinationUpdate = Time.realtimeSinceStartup;
+            var combination = CombinationFromKeyboard();
+            if (_currentCombination != null && _currentCombination != combination)
+            {
+                _currentCombination.SetPressed(false);
+            }
+            if (combination == null)
+            {
+                _currentCombination = null;
+                return;
+            }
+            _currentCombination = (ModInputCombination)combination;
+            _currentCombination.SetPressed();
+        }
+
+        public bool IsPressedExact(IModInputCombination combination)
+        {
+            UpdateCurrentCombination();
+            return _currentCombination == combination;
+        }
+
+        public bool IsNewlyPressedExact(IModInputCombination combination, bool keep = false)
+        {
+            UpdateCurrentCombination();
+            return _currentCombination == combination && combination.IsFirst(keep);
+        }
+
+        public bool WasTappedExact(IModInputCombination combination, bool keep = false)
+        {
+            UpdateCurrentCombination();
+            return combination != _currentCombination
+                && (combination.PressDuration < TapDuration)
+                && combination.IsFirst(keep);
+        }
+
+        public bool WasNewlyReleasedExact(IModInputCombination combination, bool keep = false)
+        {
+            UpdateCurrentCombination();
+            return _currentCombination != combination && combination.IsFirst(keep);
+        }
+
+        private void UpdateSinglesPressed()
+        {
+            if (!(Time.realtimeSinceStartup - _lastSingleUpdate > Time.unscaledDeltaTime / 2))
+            {
+                return;
+            }
+            _lastSingleUpdate = Time.realtimeSinceStartup;
+            var toRemove = new List<IModInputCombination>();
+            foreach (var combo in _singlesPressed)
+            {
+                if (!IsPressed(combo))
+                {
+                    ((ModInputCombination)combo).SetPressed(false);
```

Comment:
This casting is an issue. I notice it's done many places.
Consider if another implementation of the interface is used. It should be safe to swap implementations of interfaces, but here it isn't. Other implementations probably won't have `.SetPressed(false);`, for example.
I guess it's best to add all of these methods to the interfaces and remove the casting.

Reply:
I'm not sure how to mark these internal functions as not to be used by mods or if mods can access internal level interfaces without being able to use them

Reply:
Actually, it seems like internal access modifier doesn't let things be used even it other project

Reply:
well, one thing I can do is add IModInputCombinationInternal interface to OWML,ModHelper.Input project and make functions only accept ModInputCombination instead of IModInputCombination

Reply:
since this class is easily accessable by mods either way

Reply:
now internal methods are public, and I marked SetPressed as Obsolete("For internal use")